chore(RHTAPWATCH-568): Add authentication to Service and ServiceMonitor

Kousalya1998 opened the pull request:
- Add authentication to Service and ServiceMonitor

Kousalya1998 commented:
@hmariset @yftacherzog @avi-biton PTAL at this PR.

yftacherzog commented:
> @hmariset @yftacherzog @avi-biton PTAL at this PR. 

Did you confirm it to be working as before @Kousalya1998

yftacherzog commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -66,9 +83,14 @@ spec:
       - name: grafana-datasource-exporter
         image: quay.io/redhat-appstudio/o11y-prometheus-exporters:v0.1
+        args:
+        - "--secure-listen-address=0.0.0.0:8443"
+        - "--upstream=http://127.0.0.1:8081/"
+        - "--logtostderr=true"
+        - "--v=10"
```
Are those arguments that our image uses?

Kousalya1998 replied:
removed

yftacherzog commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -10,14 +10,31 @@ metadata:
-- apiGroups: ['grafana.integreatly.org']
-  resources: ['grafanas']
-  verbs: ["get"]
+# - apiGroups: ['grafana.integreatly.org']
+#   resources: ['grafanas']
+#   verbs: ["get"]
```
Can we delete the commented lines?

Kousalya1998 replied:
done

avi-biton commented on exporters/dsexporter/dsexporter.go:
```diff
@@ -43,13 +43,14 @@ func main() {
 	))
-	fmt.Println("Server is listening on http://localhost:8090/metrics")
-	http.ListenAndServe(":8090", nil)
-}
```
Missing a new line

Kousalya1998 replied:
@avi-biton updating it now. I was troubleshooting the code and with https it gives me the merics for port 8443. Any idea which port should I be using here?

avi-biton replied:
You may define whichever port you want, 8443 seems like a good choice :)

avi-biton replied:
Once it will be defined and deployed, we should be able to see it in the `targets` in openshift

Kousalya1998 replied:
can see the targets but it says scrape failed. Ports are defined correctly 
![image](https://github.com/redhat-appstudio/o11y/assets/113047440/8c9bdd38-5f47-437c-9249-9dd6dc7ce2f4)

avi-biton commented:
@Kousalya1998 were you able to check this configuration?

avi-biton commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -88,8 +99,13 @@ metadata:
   - path: /metrics
-    port: http
-    scheme: http
+    port: "8090"
```
```suggestion
    port: https
```

Kousalya1998 replied:
done.

yftacherzog reviewed (changes requested):
There are still some pieces missing to make it work. See my comments here: https://redhat-internal.slack.com/archives/C04FDFTF8EB/p1706166238412489?thread_ts=1706163051.438089&cid=C04FDFTF8EB

avi-biton commented on config/exporters/monitoring/grafana/base/prometheus-dsexporter-service.yaml:
```diff
@@ -0,0 +1,112 @@
+      containers:
+      - name: grafana-datasource-exporter
+        image: exporter:latest
+        image: quay.io/redhat-appstudio/o11y-prometheus-exporters:v0.1
```
One of the images should be removed

avi-biton commented on config/exporters/monitoring/grafana/base/prometheus-dsexporter-service.yaml:
```diff
@@ -0,0 +1,112 @@
+    port: https
+    scheme: https
+    bearerTokenSecret:
+      name: "availability-exporter-sa"
```
We are using the service account that was defined and used in the deployment.
We should create a new service account for the servicemonitor

Kousalya1998 replied:
@avi-biton So, now the new service account will be used for the Secret and Service monitor?
Service & Deployment uses the `exporter-sa` service account

yftacherzog commented:
Is it working now? If so, are you no longer able to reach the /metrics endpoint by port-forwarding the service?
Per the example in the ticket and in the dummy service monitor, I was expecting we will need the rbac-proxy container to be deployed as well here, otherwise, I'm not sure what's implementing the authentication.

Kousalya1998 commented:
> Is it working now? If so, are you no longer able to reach the /metrics endpoint by port-forwarding the service? Per the example in the ticket and in the dummy service monitor, I was expecting we will need the rbac-proxy container to be deployed as well here, otherwise, I'm not sure what's implementing the authentication.

I am not able to see the metrics using port-forwarding now. 

Yes, I made the modifications that Avi suggested. I also referred to the file you shared with me, where they are asking to pass the  `--upstream flag` as an argument in order for it to perform the authorization. Which is throwing some error when I use those args.

ref doc: https://github.com/brancz/kube-rbac-proxy/tree/master/examples/non-resource-url 

Also, I noticed that the same example uses a Dockerfile which configures the container to make an HTTP request to a Kubernetes RBAC proxy service to retrieve metrics using an authorization token from the Kubernetes service account.

cc: @yftacherzog @hmariset @avi-biton 

Slack thread: https://redhat-internal.slack.com/archives/C04FDFTF8EB/p1706163051438089

avi-biton commented:
When we are using the `kube-rbac-proxy`, we have to configure the communication between our service to the `kube-rbac-proxy`.
By adding the `- "--upstream=http://127.0.0.1:8081/"` to the `kube-rbac-proxy` and adding an argument to our service:
```
        args:
        - "--bind=127.0.0.1:8081"
```
We make this connection between the 2.
However, I suspect the `o11y-prometheus-exporters` is not able to receive arguments as it is constructed now.

hmariset commented on Dockerfile:
```diff
@@ -7,15 +7,14 @@ COPY go.mod go.sum ./
 
-FROM registry.access.redhat.com/ubi9/ubi-minimal:9.3-1361.1699548032
-RUN microdnf update --setopt=install_weak_deps=0 -y && microdnf install -y libcurl-minimal libcurl-devel
+FROM quay.io/prometheus/busybox:latest
```
Please update this image to redhat image.

hmariset commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service-monitor.yaml:
```diff
@@ -0,0 +1,56 @@
+kind: ServiceAccount
+metadata:
+  name: metrics-reader
+  namespace: dummy-service-test
```
update the namespace value

hmariset commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -1,95 +1,108 @@
 kind: Namespace
 metadata:
-  name: appstudio-grafana-datasource-exporter
+  name: dummy-service-test
```
this change is not required, you can name it appstudio-grafana-datasource-exporter

yftacherzog replied:
@Kousalya1998 there are still multiple mentions of `dummy` inside the yamls.
Please rename everything so it reflects its use instead of dummy-something.

hmariset reviewed (changes requested):
rebase to updated exporters code, update namespace values and image used in Dockerfile.

Kousalya1998 commented:
> rebase to updated exporters code, update namespace values and image used in Dockerfile.

@hmariset I rebased and updated the exporter code, namespaces, and image used in the Dockerfile.

Kousalya1998 commented:
@amisstea @avi-biton @yftacherzog @hmariset PTAL at the updated PR

yftacherzog commented:
> @amisstea @avi-biton @yftacherzog @hmariset PTAL at the updated PR

@Kousalya1998 unit tests are failing

Kousalya1998 commented:
@yftacherzog now all unit test case passes

yftacherzog commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service-monitor.yaml:
```diff
@@ -0,0 +1,56 @@
+type: kubernetes.io/service-account-token
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
```
I think we should make it a role rather than a cluster role, and the binding to be a role binding. We're staying within the namespace as far as I see it. No need to potentially allow access outside of the namespace.

yftacherzog replied:
This question still stands

yftacherzog commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -7,89 +9,100 @@ spec: {}
-- apiGroups: ['grafana.integreatly.org']
-  resources: ['grafanas']
-  verbs: ["get"]
+- apiGroups: ["authentication.k8s.io"]
+  resources:
+  - tokenreviews
+  verbs: ["create"]
+- apiGroups: ["authorization.k8s.io"]
+  resources:
+  - subjectaccessreviews
+  verbs: ["create"]
```
What's the reason for changing this? Why do we need `create` now? Why don't we need `get` anymore?

yftacherzog commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -7,89 +9,100 @@ spec: {}
 ---
 apiVersion: rbac.authorization.k8s.io/v1
-kind: RoleBinding
+kind: ClusterRoleBinding
```
Why is this changing?

Kousalya1998 replied:
Updated

yftacherzog commented on exporters/dsexporter.go:
```diff
@@ -77,13 +79,12 @@ func GetGrafanaResource(clientset *kubernetes.Clientset) (map[string]interface{}
 
 // get datasources from grafana resource
 func GetDataSources(grafanaResource map[string]interface{}) ([]string, error) {
-	// return empty string slice if datasources are not defined
```
Is this change related to this PR?

yftacherzog commented on exporters/dsexporter.go:
```diff
@@ -120,6 +125,6 @@ func main() {
 
-	fmt.Println("Server is listening on http://localhost:8090/metrics")
-	http.ListenAndServe(":8090", nil)
+	fmt.Printf("Server is listening on %s/metrics\n", bind)
```
Are we good with what it prints now? It's not the same as before.

yftacherzog reviewed (changes requested):
@Kousalya1998 I still think this PR requires quite a lot of cleanup.

- Make sure the changes you make are the ones required to fulfill the task.
- Make sure naming makes sense - we're starting to finalize this code. We should not use placeholder anymore.
- This task is about security. Make sure we provide just enough access and not more than what is needed.
- Once it's ready for review, please deploy it in dev using infra-deployments' preview script and verify everything still works as expected.

hmariset commented on Dockerfile:
```diff
@@ -7,15 +7,14 @@ COPY go.mod go.sum ./
 
 
-FROM registry.access.redhat.com/ubi9/ubi-minimal:9.3-1361.1699548032
-RUN microdnf update --setopt=install_weak_deps=0 -y && microdnf install -y libcurl-minimal libcurl-devel
```
To remove vulnerabilities during security scan I added line 18 but with ubi-micro image this is not required.

Kousalya1998 commented:
@yftacherzog 
I now chnaged the ones required to fulfill the task
Removed the placeholders and followed proper naming format
deployed it in dev using infra-deployments' preview script and verified everything works as expected.

yftacherzog commented on config/exporters/monitoring/grafana/base/kustomization.yaml:
```diff
@@ -1,8 +1,8 @@
 resources:
+- prometheus-exporter-service-monitor.yaml
 - prometheus-exporter-service.yaml
-images:
-- name: exporter
-  newName: quay.io/redhat-appstudio/o11y
-  newTag: latest
```
What's the reason for this change?

Kousalya1998 replied:
I removed it thinking newTag was not being used. reverted the changed and renamed the image to exporter:latest

yftacherzog commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -7,7 +9,7 @@ spec: {}
 kind: ServiceAccount
 metadata:
-  name: exporter-sa
+  name: kube-rbac-proxy
```
What's the reason for renaming? As far as I understand, this is the service account that allows us to access the grafana datasource. I would not expect it to change in this task. I'd only expect that service account that allows prometheus to read from the exporter to be affected here.

Kousalya1998 replied:
I tried reverting to `exporter-sa` and its throwing gin a 401 forbidden error.

yftacherzog replied:
In previous PRs, we changed the code so that the exporter is reading data from grafana. A service account was required for that. I believe this is that service account.
Now we want to make the exporter itself secured and for that we need another service account.

Those two service accounts should be unrelated. We should not change the first one to make the second one work. Not sure what is throwing the error now, but it should not be related.

Kousalya1998 replied:
@yftacherzog 
I added a new service account & I tried changing the serviceaccount within the Deployment to `exporter-sa` and the Target is  down and not being scraped for metrics.

Next, I left the `serviceAccountName` within the `Deployment` to remain as `kube-rbac-proxy`, that way the two service accounts `exporter-sa` & `kube-rbac-proxy` (kube-rbac-proxy for reading data from Grafana) remains unrelated. After deploying it in dev, the target is currently up and being actively scraped for metrics.

yftacherzog commented on exporters/dsexporter.go:
```diff
@@ -107,6 +109,10 @@ func GetDataSourceAvailability(datasources []string, dsToCheck string) float64 {
 // providing a cleaner and more focused view. Also reduces potential noise from unrelated metrics.
 func main() {
+	var bind string
+	flag.StringVar(&bind, "bind", ":8090", "The socket to bind to.")
```
The value that we're providing to `bind` on the yaml is a combination of the address and the port, but the default value we give here contains only the port. Does it make sense?

Kousalya1998 replied:
I removed the bind and the args from yaml. I think the args are not being used since we are directly using in the port to be 8090. Also, I checked that its being authenticated. No issues in removing the args.

hmariset commented on config/exporters/monitoring/grafana/base/prometheus-exporter-service.yaml:
```diff
@@ -22,74 +25,78 @@ rules:
+          readOnlyRootFilesystem: true
+          runAsNonRoot: true
+      - name: exporters
+        image: quay.io/klakshma/myfirstrepo
```
update the image to `quay.io/redhat-appstudio/o11y`

Kousalya1998 replied:
updated

hmariset commented on exporters/dsexporter_test.go:
```diff
@@ -133,5 +133,4 @@ func TestMain(t *testing.T) {
 	assert.Equal(t, float64(0), testutil.ToFloat64(exporter))
 
 	allDataSources = MockGetDataSourcesExists
-	assert.Equal(t, float64(1), testutil.ToFloat64(exporter))
```
Why is this line removd?

Kousalya1998 replied:
reverted changes.

Kousalya1998 commented on exporters/dsexporter_test.go:
```diff
@@ -129,9 +129,20 @@ func TestMain(t *testing.T) {
 
 	assert.Equal(t, http.StatusOK, res.StatusCode)
-	assert.Equal(t, 1, testutil.CollectAndCount(exporter))
-	assert.Equal(t, float64(0), testutil.ToFloat64(exporter))
```
reverted changes.

hmariset reviewed (commented):
Check for yaml linting errors. Everything else looks good for me.

Kousalya1998 commented:
> Check for yaml linting errors. Everything else looks good for me.

yes checked

yftacherzog commented on Dockerfile:
```diff
@@ -7,15 +7,14 @@ COPY go.mod go.sum ./
 # Download required dependencies
 RUN go mod download
-COPY --chown=default:root exporters/**/*.go .
+COPY --chown=default:root exporters/*.go .
```
Do we need this change?

Kousalya1998 replied:
reverted the changes

yftacherzog commented on Dockerfile:
```diff
@@ -28,4 +27,4 @@ LABEL io.k8s.display-name="o11y-exporters"
 LABEL summary="Konflux Observability Exporters"
 
-CMD ["/bin/exporters"]
+ENTRYPOINT ["/bin/exporters"]
```
We can probably revert this as we're not using args for now

Kousalya1998 replied:
updated

yftacherzog commented on Dockerfile:
As mentioned on the call, assuming that we don't need for this PR any changes other than the ones under `config`, I think we should revert all changes here and in the go code.

hmariset commented on Dockerfile:
```diff
@@ -28,4 +27,4 @@ LABEL io.k8s.display-name="o11y-exporters"
 LABEL summary="Konflux Observability Exporters"
 
-CMD ["/bin/exporters"]
+CMD ["/bin/exporters"]
```
new line missing at end of the file

hmariset reviewed (approved):
Tested the code and the target is up now.

yftacherzog reviewed (changes requested):
See my comment from yesterday

avi-biton commented on exporters/dsexporter/dsexporter_test.go:
```diff
@@ -134,4 +134,4 @@ func TestMain(t *testing.T) {
 	allDataSources = MockGetDataSourcesExists
 	assert.Equal(t, float64(1), testutil.ToFloat64(exporter))
-}
+}
```
Please add new line at the end

Kousalya1998 replied:
reverted the changes I made

yftacherzog commented:
Please confirm that you're able to deploy it with infra-deployment's preview script and it works as expected.

Kousalya1998 commented:
> Please confirm that you're able to deploy it with infra-deployment's preview script and it works as expected.

@yftacherzog Just deployed it with the infra-deployments preview script and it works as expected

yftacherzog reviewed (approved):
lgtm
